transformer_engine/jax/praxis/module.py

```diff
@@ -0,0 +1,322 @@
+# Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+"""
+Praxis Modules
+"""
+from functools import partial
+from typing import Callable, Iterable, Sequence, Tuple, Union
+
+import jax.numpy as jnp
+from praxis.base_layer import init_var
+from praxis.base_layer import BaseLayer, WeightInit, WeightHParams
+from praxis.layers import flax_adapter
+from praxis.pytypes import JTensor
+
+from ..module import DenseGeneral, LayerNormDenseGeneral
+from ..module import LayerNorm as flax_LayerNorm
+from ..module import LayerNormMLP as flax_LayerNormMLP
+from ..module import Softmax
+from ..softmax import SoftmaxType
+from ..sharding import MajorShardingType, ShardingType
+
+
+class TransformerEngineBaseLayer(BaseLayer):
+    """TransformerEngineBaseLayer"""
+
+    class HParams(BaseLayer.HParams):    # pylint: disable=too-few-public-methods
+        """TransformerEngineBaseLayer.HParams"""
+        logical_axes_rules: Tuple[Tuple, ...] = None
+
+    @staticmethod
+    def generate_params_init(name: str, initializer: WeightInit):
+        """generate_params_init"""
+
+        def kernel_init(key, shape, dtype):
+            wp = WeightHParams(shape=shape, init=initializer, dtype=dtype)
+            return init_var(name, wp, key)
+
+        return kernel_init
+
+    def create_layer(self, name, flax_module_cls, module_init_args_functor,
+                     module_init_kwargs_functor):
+        """create_layer"""
+        flax_module_p = flax_adapter.FlaxModuleAdapter.HParams(
+            module_factory_method=flax_module_cls,
+            var_init_args=module_init_args_functor,
+            var_init_kwargs=module_init_kwargs_functor,
+            logical_axes_rules=self.hparams.logical_axes_rules,
+            ici_mesh_shape=self.hparams.ici_mesh_shape,
+            dcn_mesh_shape=self.hparams.dcn_mesh_shape,
+            mesh_axis_names=self.hparams.mesh_axis_names)
+
+        self.create_child(name, flax_module_p.clone())
```

Comment:
Why clone?

Reply:
It is a kind of suggestion of Praxis to avoid unknown pollution to `flax_module_p`, in case we need some information of `flax_module_p` later.

Reply:
If we retain flax_module_p, I agree it can be useful. But here we don't keep it. The original object will be deleted.
Did you intend to keep a copy of it?

Reply:
I have no strong reason to keep that copying here. We did that just to align the coding style as Praxis did. Will remove it in the official release.

Reply:
I don't have a strong opinion here, just wanted to understand.